### PR TITLE
Tooling deployment scaffolding 

### PR DESCRIPTION
**Pull request by connorwstein**
- NOTE ~5k is generated protobuf code to unblock. That will be imported once exposed. 
- We put the deployment/configuration logic in integration-tests module for a few reasons:
	- Keeps the chain dependencies out of the core module, in particular helpful for eventual cross family tests
	- It can become the canonical deployment logic to be used for CRIB envs as well (eventually can replace the actions + contracts dirs)  
- To accomplish the lightweight tests (chainlink.Application + simulated.Backend) we expose some test utilities in util/testutils/
- integration-tests/deployment holds product agnostic deployment utilities including a general purpose environment structure to write environment abstracted code against and migration output components (address books, proposals etc)
- integration-tests/deployment/ccip holds all product specific deployment code including
	- Top level migrations and migration tests where a "migration" is defined to be a function which operates against an environment and outputs a MigrationOutput structure with one or more artifacts (MCMS proposals, job specs). Notably migration tests can apply those outputs to an ephemeral environment to ensure correctness. These migrations are intended for export and use against real environments (testnet/mainnet).
	- Re-usable product specific components of top level migrations and associated tests

Next steps / follow up PRs:
- Port testutils export to chainlink repo
- Example solana setup
- Once cross family validated, start deeper testing and real CCIP use cases

### CODE REVIEW

File: integration-tests/deployment/memory/node.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+type Node struct {
+	App chainlink.Application
+	// Transmitter key/OCR keys for this node
+	Keys       Keys
+	Addr       net.TCPAddr
+	IsBoostrap bool
+}
+
+type RegistryConfig struct {
```
Comment (AnieeG) [inline integration-tests/deployment/memory/node.go:64]: Is this going to be common for all? If yes, can we port it to a `types` package?

Comment (connorwstein) [reply integration-tests/deployment/memory/node.go:64]: we can pull it up to deployment pkg I think yeah (that is functioning like a types package)

File: integration-tests/deployment/environment.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		selectors = append(selectors, sel)
+	}
+	return selectors
+}
+
+func ConfirmIfNoError(chain Chain, tx *types.Transaction, err error) error {
+	if err != nil {
+		d, ok := err.(rpc.DataError)
+		if ok {
+			fmt.Println("Got Data Error", d.ErrorData())
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go]: would help with a logger

File: integration-tests/deployment/environment.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+func MaybeDataErr(err error) string {
+	d, ok := err.(rpc.DataError)
+	if ok {
+		return d.ErrorData().(string)
+	}
+	return ""
+}
+
+func UBigInt(i uint64) *big.Int {
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:85]: how about moving to a utils/helper ?

File: integration-tests/deployment/environment.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	return new(big.Int).SetUint64(i)
+}
+
+func E18Mult(amount uint64) *big.Int {
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:89]: how about moving to a utils/helper ?

File: integration-tests/deployment/environment.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	return peerIDs
+}
+
+// OffchainPublicKey types.OffchainPublicKey
+// // For EVM-chains, this an *address*.
+// OnchainPublicKey types.OnchainPublicKey
+// PeerID           string
+// TransmitAccount  types.Account
+type Node struct {
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:134]:
```suggestion
type OCRNode struct {
```

File: integration-tests/deployment/environment.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	nodev1 "github.com/smartcontractkit/chainlink/integration-tests/deployment/jd/node/v1"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/p2pkey"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+type OnchainClient interface {
+	// For EVM specifically we can use existing geth interface
+	// to abstract chain clients.
+	bind.ContractBackend
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:28]: can we also include `bind.DeployBackend` ?

File: integration-tests/deployment/environment.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	// Note the Sign function can be abstract supporting a variety of key storage mechanisms (e.g. KMS etc).
+	DeployerKey *bind.TransactOpts
+	Confirm     func(tx common.Hash) error
+}
+
+type Environment struct {
+	Name     string
+	Chains   map[uint64]Chain
+	Offchain OffchainClient
+	NodeIDs  []string
```
Comment (AnieeG) [inline integration-tests/deployment/environment.go:50]: Are these NodeIDs specific to JD ?

File: integration-tests/deployment/memory/environment.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+		chains[sel] = deployment.Chain{
+			Selector:    sel,
+			Client:      chain.Backend,
+			DeployerKey: chain.DeployerKey,
+			Confirm: func(tx common.Hash) error {
+				for {
+					chain.Backend.Commit()
+					receipt, err := chain.Backend.TransactionReceipt(nil, tx)
+					if err != nil {
+						fmt.Println("failed to get receipt", err)
```
Comment (AnieeG) [inline integration-tests/deployment/memory/environment.go]: A logger would be more useful

File: integration-tests/deployment/address_book.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
+package deployment
+
+import "fmt"
+
+type AddressBook interface {
+	Save(chainSelector uint64, address string, typeAndVersion string) error
+	Addresses() (map[uint64]map[string]string, error)
```
Comment (makramkd) [inline integration-tests/deployment/address_book.go:9]: Having comments on what the keys on this map mean might help. I'm guessing its chainSelector => (typeAndVersion => address) but not 100% sure.

Alternatively a type alias on ChainSelector/Address and maybe a concrete TypeAndVersion type e.g

```golang
type (
  ChainSelector = uint64
  TypeAndVersion struct {
    Type, Version string
  }
  Address = string
)
```

File: integration-tests/deployment/address_book.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	m.AddressesByChain[chainSelector][address] = typeAndVersion
+	return nil
+}
+
+func (m *AddressBookMap) Addresses() (map[uint64]map[string]string, error) {
+	return m.AddressesByChain, nil
+}
+
+func (m *AddressBookMap) AddressesForChain(chain uint64) (map[string]string, error) {
+	// TODO error
```
Comment (makramkd) [inline integration-tests/deployment/address_book.go]: To implement?

File: integration-tests/deployment/address_book.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+// Attention this will mutate existing book
+func (m *AddressBookMap) Merge(ab AddressBook) error {
+	addresses, err := ab.Addresses()
+	if err != nil {
+		return err
+	}
+	for chain, chainAddresses := range addresses {
+		for address, typeAndVersions := range chainAddresses {
+			return m.Save(chain, address, typeAndVersions)
```
Comment (makramkd) [inline integration-tests/deployment/address_book.go]: Seems like a premature return here, is this intended?

File: integration-tests/deployment/address_book.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
+package deployment
+
+import "fmt"
+
+type AddressBook interface {
```
Comment (makramkd) [inline integration-tests/deployment/address_book.go:7]: Doc comment on this would be useful to explain significance and also to indicate whether implementations should be thread-safe

File: integration-tests/deployment/address_book.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	AddressesForChain(chain uint64) (map[string]string, error)
+	// Allows for merging address books (e.g. new deployments with existing ones)
+	Merge(other AddressBook) error
+}
+
+type AddressBookMap struct {
+	AddressesByChain map[uint64]map[string]string
+}
+
+func (m *AddressBookMap) Save(chainSelector uint64, address string, typeAndVersion string) error {
```
Comment (makramkd) [inline integration-tests/deployment/address_book.go:19]: Should we check for edge cases here e.g address or typeAndVersion being empty strings?

File: integration-tests/deployment/ccip/deploy.go [hunk 14]
```diff
@@ -0,0 +1,565 @@
...
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/price_registry"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/router"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/token_admin_registry"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/weth9"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/capabilities_registry"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/shared/generated/burn_mint_erc677"
+)
+
+var (
+	// 1.0
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go:32]: I wonder if it would be better to programmatically generate this list somehow so that we don't miss any contracts.

File: integration-tests/deployment/ccip/deploy.go [hunk 15]
```diff
@@ -0,0 +1,565 @@
...
+		*owner_helpers.ManyChainMultiSig |
+		*owner_helpers.RBACTimelock |
+		*evm_2_evm_multi_offramp.EVM2EVMMultiOffRamp |
+		*evm_2_evm_multi_onramp.EVM2EVMMultiOnRamp |
+		*burn_mint_erc677.BurnMintERC677 |
+		*maybe_revert_message_receiver.MaybeRevertMessageReceiver
+}
+
+type ContractDeploy[C Contracts] struct {
+	// We just return keep all the deploy return values
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go]:
```suggestion
	// We just keep all the deploy return values
```

File: integration-tests/deployment/ccip/deploy.go [hunk 16]
```diff
@@ -0,0 +1,565 @@
...
+	// since some will be empty if there's an error.
+	// and we want to avoid repeating that
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go]: Avoid repeating what? The deployment?

File: integration-tests/deployment/ccip/deploy.go [hunk 17]
```diff
@@ -0,0 +1,565 @@
...
+	Address  common.Address
+	Contract C
+	Tx       *types.Transaction
+	TvStr    string
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go:78]: I think you can use `TypeAndVersion` here since its a string in the address book too right?

File: integration-tests/deployment/ccip/deploy.go [hunk 18]
```diff
@@ -0,0 +1,565 @@
...
+	if err := AddNodes(
+		c.CapabilityRegistry[c.HomeChainSel],
+		e.Chains[c.HomeChainSel],
+		nodes.PeerIDs(c.HomeChainSel), // Doesn't actually matter which sel here
+		[][32]byte{cap},
+	); err != nil {
+		return ab, err
+	}
+
+	for sel, chain := range e.Chains {
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go:142]: Loop is kinda long and hard to follow, might be simpler if its broken up into funcs that do specific deploys / setups

File: integration-tests/deployment/ccip/deploy.go [hunk 19]
```diff
@@ -0,0 +1,565 @@
...
+		); err != nil {
+			e.Logger.Errorw("Failed to add DON", "err", err)
+			return ab, err
+		}
+	}
+
+	return ab, nil
+}
+
+func AddLane(e deployment.Environment, state CCIPOnChainState, from, to uint64) error {
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy.go:457]: Similarly here maybe break up into logical pieces that can fit into functions

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+	for _, pluginType := range []cctypes.PluginType{cctypes.PluginTypeCCIPCommit, cctypes.PluginTypeCCIPExec} {
+		_, err = offRamp.LatestConfigDetails(&bind.CallOpts{
+			Context: context.Background(),
+		}, uint8(pluginType))
+		if err != nil {
+			//return err
+			return fmt.Errorf("%s", err.(rpc.DataError).ErrorData().(string))
+		}
+		// TODO: assertions
+		//require.Equalf(t, offrampOCR3Configs[pluginType].ConfigDigest, ocrConfig.ConfigInfo.ConfigDigest, "%s OCR3 config digest mismatch", pluginType.String())
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go:398]: Were these assertions failing? Or is it just a matter of not using `require`?

File: integration-tests/deployment/ccip/deploy_home_chain.go [hunk 21]
```diff
@@ -0,0 +1,416 @@
...
+		return err
+	}
+
+	for _, pluginType := range []cctypes.PluginType{cctypes.PluginTypeCCIPCommit, cctypes.PluginTypeCCIPExec} {
+		_, err = offRamp.LatestConfigDetails(&bind.CallOpts{
+			Context: context.Background(),
+		}, uint8(pluginType))
+		if err != nil {
+			//return err
+			return fmt.Errorf("%s", err.(rpc.DataError).ErrorData().(string))
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_home_chain.go]: Cast here always safe? I think you had a helper function for this somewhere in the PR

File: integration-tests/deployment/ccip/jobs.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+			NodeId: node,
+		}})
+		if err != nil {
+			return nil, err
+		}
+		for _, chainConfig := range nodeChainConfigs.ChainConfigs {
+			if chainConfig.Ocr2Config.IsBootstrap {
+				bootstrapMp[fmt.Sprintf("%s@%s",
+					// p2p_12D3... -> 12D3...
+					chainConfig.Ocr2Config.P2PKeyBundle.PeerId[4:], chainConfig.Ocr2Config.Multiaddr)] = struct{}{}
```
Comment (makramkd) [inline integration-tests/deployment/ccip/jobs.go:34]: `Multiaddr` seems like a poor name, what does multi mean in this instance? Seems like its the p2p listen address but not 100% sure

File: integration-tests/deployment/ccip/jobs.go [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+			}
+		}
+	}
+	var bootstraps []string
+	for b := range bootstrapMp {
+		bootstraps = append(bootstraps, b)
```
Comment (makramkd) [inline integration-tests/deployment/ccip/jobs.go:40]: q: any reason you can't just append in the loop above? Is there a risk that multiple nodes have the same p2p ID and multiaddr? Should be impossible right?

File: integration-tests/deployment/ccip/jobs.go [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+		}})
+		if err != nil {
+			return nil, err
+		}
+		spec, err := validate.NewCCIPSpecToml(validate.SpecArgs{
+			P2PV2Bootstrappers:     bootstraps,
+			CapabilityVersion:      CapabilityVersion,
+			CapabilityLabelledName: CapabilityLabelledName,
+			OCRKeyBundleIDs: map[string]string{
+				// TODO: Validate that that all EVM chains are using the same keybundle.
```
Comment (makramkd) [inline integration-tests/deployment/ccip/jobs.go:56]: Does this validation refer to the ocr keys specified in the onchain ocr config? Or somewhere else?

File: integration-tests/deployment/ccip/migrations/1_initial_deploy.go [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment"
+
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip"
+)
+
+// We expect the migration input to be unique per migration.
+// TODO: Maybe there's a generics approach here?
+// Note if the migration is a deployment and it fails we have 2 options:
+// - Just throw away the addresses, fix issue and try again (potentially expensive on mainnet)
+// - Roll forward with another migration completing the deployment
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy.go:13]: Yeah I think having a "resume" for a migration would be great if it can pick up from the last successful deployment. Tools like Chef do this ("idempotency")

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+		}
+	}
+
+	// Send a message from each chain to the home chain.
+	for sel, chain := range e.Chains {
+		dest := homeChainSel
+		if sel == homeChainSel {
+			continue
+		}
+		//require.NoError(t, ccipdeployment.AddLane(e, state, sel, dest))
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: Can delete line?

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 27]
```diff
@@ -0,0 +1,267 @@
...
+	// Add all lanes
+	for source := range e.Chains {
+		for dest := range e.Chains {
+			if source != dest {
+				require.NoError(t, ccipdeployment.AddLane(e, state, source, dest))
+			}
+		}
+	}
+
+	// Send a message from each chain to the home chain.
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: Can we make the test like the current test where we send to every other chain? It would test the multi-source batching

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 28]
```diff
@@ -0,0 +1,267 @@
...
+	}, sink)
+	require.NoError(t, err)
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-time.After(time.Minute):
+			t.Logf("Waiting for commit report on chain selector %d from source selector %d expected seq nr range %s",
+				dest.Selector, src.Selector, expectedSeqNumRange.String())
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: Move it to the `<-ticker.C` case?

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 29]
```diff
@@ -0,0 +1,267 @@
...
+			t.Error("Timed out waiting for commit report")
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: I dunno if you need this kinda logic, feels like it makes the test more brittle. Why not watch for all commit reports and have the test time out do the timing out like we do now?

File: integration-tests/deployment/memory/environment.go (hunk 8 above)
Comment (makramkd) [reply integration-tests/deployment/memory/environment.go]: Or just `t.Log`

**Review by makramkd — COMMENTED** [review]
PR is kinda huge so hard to review the whole thing but looks good overall. Have some comments about the test logic changes, not sure it makes the test more robust

File: integration-tests/deployment/address_book.go [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+	if _, exists := m.AddressesByChain[chainSelector]; !exists {
+		// First time chain add, create map
+		m.AddressesByChain[chainSelector] = make(map[string]string)
+	}
+	if _, exists := m.AddressesByChain[chainSelector][address]; exists {
+		return fmt.Errorf("address %s already exists for chain %d", address, chainSelector)
```
Comment (asoliman92) [inline integration-tests/deployment/address_book.go:25]: It's okay if it has same value not to error, just a log should be enough.

File: integration-tests/deployment/ccip/deploy.go [hunk 31]
```diff
@@ -0,0 +1,565 @@
...
+	CapabilitiesRegistry_1_0_0 = "CapabilitiesRegistry 1.0.0"
+	CCIPConfig_1_6_0           = "CCIPConfig 1.6.0-dev"
+	EVM2EVMMultiOnRamp_1_6_0   = "EVM2EVMMultiOnRamp 1.6.0-dev"
+	EVM2EVMMultiOffRamp_1_6_0  = "EVM2EVMMultiOffRamp 1.6.0-dev"
+	NonceManager_1_6_0         = "NonceManager 1.6.0-dev"
+	PriceRegistry_1_6_0        = "PriceRegistry 1.6.0-dev"
+)
+
+type Contracts interface {
+	*capabilities_registry.CapabilitiesRegistry |
```
Comment (asoliman92) [inline integration-tests/deployment/ccip/deploy.go:55]: TIL this feature in Go :D

File: integration-tests/deployment/environment.go (hunk 3 above)
Comment (connorwstein) [reply integration-tests/deployment/environment.go:85]: I think in a follow up PR I'll dig around for a better home. Something in chainlink-common or existing tools should exist for this

File: integration-tests/deployment/environment.go (hunk 5 above)
Comment (connorwstein) [reply integration-tests/deployment/environment.go:134]: today I actually don't think there are non-OCR nodes anymore heh

File: integration-tests/deployment/environment.go (hunk 6 above)
Comment (connorwstein) [reply integration-tests/deployment/environment.go:28]: idt its needed because `Confirm` is a separate function pointer which can enclose transactionReceipt?

File: integration-tests/deployment/environment.go (hunk 7 above)
Comment (connorwstein) [reply integration-tests/deployment/environment.go:50]: right now I'm using the peerIDs to unblock but JD does have its own. Will have to move to theirs at some point or convince them to use peerIDs

File: integration-tests/deployment/address_book.go (hunk 9 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go:9]: yeah type alias sgtm. Can add in a follow up with further address book testing

File: integration-tests/deployment/address_book.go (hunk 11 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go]: nope good catch, merge to be tested in a follow up / once we determine its use case a bit better

File: integration-tests/deployment/address_book.go (hunk 12 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go:7]: yeah I think a full polish address book follow up PR needed

File: integration-tests/deployment/ccip/deploy.go (hunk 14 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/deploy.go:32]: probably possible - maybe overkill? Hard to imagine you forget a whole contract in your system

File: integration-tests/deployment/ccip/deploy.go (hunk 16 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/deploy.go]: just to avoid a zillion if err != nil {..} do sth with tx.Hash()

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 20 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/deploy_home_chain.go:398]: not using require

File: integration-tests/deployment/ccip/deploy_home_chain.go (hunk 21 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/deploy_home_chain.go]: fixed

File: integration-tests/deployment/ccip/jobs.go (hunk 22 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/jobs.go:34]: I inherited the name from JD interface. Can change them both atomically in a follow up

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go (hunk 29 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: I'm ok to remove. Default test timeout just felt really long

File: integration-tests/deployment/address_book.go (hunk 30 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go:25]: in this context I think it'd imply caller has a bug so should error

File: integration-tests/deployment/ccip/migrations/1_initial_deploy.go (hunk 25 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/migrations/1_initial_deploy.go:13]: yeah TODO

File: integration-tests/deployment/ccip/jobs.go (hunk 24 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/jobs.go:56]: offchain OCR keys on the node

File: integration-tests/deployment/ccip/jobs.go (hunk 23 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/jobs.go:40]: its because the p2pID lives in the per chain per node config, so need to dedupe across chains

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go (hunk 27 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go]: yeah was gonna come back to that, but should be easy will do it now

File: integration-tests/deployment/address_book.go (hunk 11 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go]: added a quick merge test

File: integration-tests/deployment/ccip/deploy_test.go [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+	e := memory.NewMemoryEnvironment(t, lggr, zapcore.InfoLevel, memory.MemoryEnvironmentConfig{
+		Chains: 1,
+		Nodes:  1,
+	})
+	js, err := NewCCIPJobSpecs(e.NodeIDs, e.Offchain)
+	require.NoError(t, err)
+	for node, jb := range js {
+		fmt.Println(node, jb)
+	}
+	// TODO: Add job assertions
```
Comment (makramkd) [inline integration-tests/deployment/ccip/deploy_test.go:48]: Maybe later can we have a ticket for each TODO (or group relevant TODO's together)?

File: integration-tests/deployment/ccip/jobs.go (hunk 24 above)
Comment (makramkd) [reply integration-tests/deployment/ccip/jobs.go:56]: Offchain OCR key usage for CCIP capability is by definition one per family (this is enforced by the job). That's why we require the bundle IDs to be set explicitly in the job spec in case there is more than one per family

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 33]
```diff
@@ -0,0 +1,295 @@
...
+		for dest, dstChain := range e.Chains {
+			if src == dest {
+				continue
+			}
+			srcChain := srcChain
+			dstChain := dstChain
+			wg.Add(1)
+			go func(src, dest uint64) {
+				defer wg.Done()
+				waitForCommitWithInterval(t, srcChain, dstChain, state.EvmOffRampsV160[dest], ccipocr3.SeqNumRange{1, 1})
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go:185]: nit: you can just use `src` and `dest` here since they're passed in as args to the function - those are eval'd at the `go` statement callsite unlike closure vars

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 34]
```diff
@@ -0,0 +1,295 @@
...
+		for dest, dstChain := range e.Chains {
+			if src == dest {
+				continue
+			}
+			srcChain := srcChain
+			dstChain := dstChain
+			wg.Add(1)
+			go func(src, dest uint64) {
+				defer wg.Done()
+				waitForExecWithSeqNr(t, srcChain, dstChain, state.EvmOffRampsV160[dest], 1)
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go:202]: Ditto here re: `src` and `dest`

File: integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go [hunk 35]
```diff
@@ -0,0 +1,295 @@
...
+	require.NoError(t, err)
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	//revive:disable
+	for {
+		select {
+		case <-ticker.C:
+			src.Client.(*backends.SimulatedBackend).Commit()
+			dest.Client.(*backends.SimulatedBackend).Commit()
```
Comment (makramkd) [inline integration-tests/deployment/ccip/migrations/1_initial_deploy_test.go:243]: There a reason we don't have these commits in a background goroutine?

**Review by makramkd — APPROVED** [review]
Some minor follow up comments; lets make sure we have tickets for all the follow up work (which seems to be quite a lot)

**Review by AnieeG — APPROVED** [review]
